### PR TITLE
flutter: reduce derivation size and refactor

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

This PR refactors the flutter derivation, with the main objective to reduce derivation size:

```
Before:
/nix/store/p3avfmvd0yrjpwwml7vvqbjin5vacx2k-flutter-stable-1.22.0       2084894968

After:
/nix/store/750k4z1yj5xsw7ymmwvn7cfjjzkaygzg-flutter-stable-1.22.0       1647047080
```

~~Also, bump the packages to the latest versions.~~

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by ericdallo — APPROVED** [review]
LGTM

**Comment by thiagokokada** [comment]
Tested in a local repo:

```
❯ /nix/store/750k4z1yj5xsw7ymmwvn7cfjjzkaygzg-flutter-stable-1.22.0/bin/flutter pub get
Downloading package sky_engine...                                   0.6s
Downloading flutter_patched_sdk tools...                            0.7s
Downloading flutter_patched_sdk_product tools...                    0.6s
Downloading linux-x64 tools...                                      1.3s
Downloading linux-x64/font-subset tools...                          0.6s
Running "flutter pub get" in mini-meta-repo...                      5.2s
```

**Comment by thiagokokada** [comment]
Opening this PR for review. I will try to bump flutter again after this PR is merged on master: https://github.com/NixOS/nixpkgs/pull/105311

**Review by ericdallo — APPROVED** [review]

**Comment by thiagokokada** [comment]
@SuperSandro2000 Why the tag of blocked by PR?

I will leave the bump in a separate PR.

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 108611` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>flutter (flutterPackages.stable)</li>
    <li>flutter-beta (flutterPackages.beta)</li>
    <li>flutter-dev (flutterPackages.dev)</li>
    <li>hover</li>
  </ul>
</details>

**Pull request merged**
